### PR TITLE
Add db::creation_tag and db::tag_depends_on metafunctions

### PR DESCRIPTION
**Pull request by wthrowe**
Also includes two cleanup commits related to DataBox to avoid recompilations for a second PR touching DataBox.hpp.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/DataStructures/DataBox/DataBox.hpp [hunk 1]
```diff
@@ -65,6 +65,38 @@ constexpr bool tag_is_retrievable_v =
...
+};
+}  // namespace detail
+
+/*!
+ * \ingroup DataBoxGroup
+ * \brief The tag added to \p Box referred to by \p Tag.  This
+ * resolves base tags and converts subitems to full items.
+ */
+template <typename Tag, typename Box>
+using real_tag =
```
Comment (kidder) [inline src/DataStructures/DataBox/DataBox.hpp]: If I understand what this is doing, in the type aliases in DataBox I have been calling the tags returned by this metafunction `creation_tags`.  If so can we use that nomenclature?

**Comment by wthrowe** [comment]
Not sure what the best way to handle the build failures is.  It seems that in this context `DEBUG_STATIC_ASSERT` is required to be followed by a semicolon in Debug mode and forbidden from being followed by a semicolon in Release mode.

File: src/DataStructures/DataBox/DataBox.hpp [hunk 2]
```diff
@@ -934,6 +988,26 @@ SPECTRE_ALWAYS_INLINE auto& get_mutable_reference(
...
+ * \brief Check whether the tag \p Consumer depends on the tag \p
+ * Provider in a given \p Box.
+ *
+ * This is equivalent to the question of whether changing \p Provider
+ * (through `db::mutate`, updating one of its dependencies, etc.)
+ * could change the value of `db::get<Consumer>`.  Tags depend on
+ * themselves, and an item and its subitems all depend on one another.
+ */
+template <typename Provider, typename Consumer, typename Box>
+using tag_depends_on =
```
Comment (kidder) [inline src/DataStructures/DataBox/DataBox.hpp:1004]: Not a change request, but I will probably get this backwards when I use/read it without looking up the documentation, as I was expecting the first template argument to be the Consumer tag that depends upon the Provider.

**Comment by kidder** [comment]
> Not sure what the best way to handle the build failures is. It seems that in this context `DEBUG_STATIC_ASSERT` is required to be followed by a semicolon in Debug mode and forbidden from being followed by a semicolon in Release mode.

Add a similar code block as we do for the false branch of ASSERT?

File: src/DataStructures/DataBox/DataBox.hpp (hunk 1 above)
Comment (wthrowe) [reply src/DataStructures/DataBox/DataBox.hpp]: Yes, those appear to be the same thing.  I will adopt that name.

File: src/DataStructures/DataBox/DataBox.hpp (hunk 2 above)
Comment (wthrowe) [reply src/DataStructures/DataBox/DataBox.hpp:1004]: I copied the order from `tt::is_a<Template, Type>`, but `std::is_base_of<Base, Derived>` is the opposite, so yeah, not obvious which is better.

**Comment by wthrowe** [comment]
Right.  Of course.  Added a commit for that (at the beginning).

Did the rename without a fixup, because it was a find-replace rename.  Waiting to see if anyone else wants to weigh in on argument ordering.

**Comment by kidder** [comment]
okay sounds good

File: src/DataStructures/DataBox/DataBox.hpp (hunk 2 above)
Comment (nilsdeppe) [reply src/DataStructures/DataBox/DataBox.hpp:1004]: The choice on `is_a` is the same as `is_base_of` in my opinion. It is `<generic, specific>`. That is, the generic template in `is_a` plays a similar role to the generic base class in `is_base_of`. That's why the ordering is that way :shrug:

Comment (wthrowe) [reply src/DataStructures/DataBox/DataBox.hpp:1004]: The English is backward, though: "X is base of Y" and "Y is a X".

Comment (nilsdeppe) [reply src/DataStructures/DataBox/DataBox.hpp:1004]: That's fair. I have no preference for what you want to do in this PR :) Just wanted to give context as to why I made `is_a` the way it is!

Comment (kidder) [reply src/DataStructures/DataBox/DataBox.hpp:1004]: If neither of you have a preference, I would prefer to reverse the order of the template parameters

Comment (nilsdeppe) [reply src/DataStructures/DataBox/DataBox.hpp:1004]: That's fine by me :)

**Review by kidder — APPROVED** [review]

**Pull request merged**